File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()
  def hey(input) do
    trimmed = String.trim(input)

    cond do
      trimmed == "" -> "Fine. Be that way!"
      is_yelling?(trimmed) && is_question?(trimmed) -> "Calm down, I know what I'm doing!"
      is_yelling?(trimmed) -> "Whoa, chill out!"
      is_question?(trimmed) -> "Sure."
      true -> "Whatever."
    end
  end

  defp is_question?(input) do
    String.ends_with?(input, "?")
  end

  defp is_yelling?(input) do
    has_letters = String.match?(input, ~r/\p{L}/u)
    has_letters && input == String.upcase(input)
  end
end

File: solutions/ruby/bob/1/bob.rb
class Bob
  RESPONSES = {
    shouting: 'Whoa, chill out!',
    question: 'Sure.',
    shouted_question: "Calm down, I know what I'm doing!",
    silent: 'Fine. Be that way!',
    default: 'Whatever.'
  }.freeze

  def self.hey(message)
    msg = message.strip

    if shouting?(msg) && question?(msg)
      respond :shouted_question
    elsif shouting?(msg)
      respond :shouting
    elsif question?(msg)
      respond :question
    elsif silent?(msg)
      respond :silent
    else
      respond :default
    end
  end

  def self.shouting?(message)
    message == message.upcase && message.match(/[a-zA-Z]/)
  end

  def self.question?(message)
    message.end_with?('?')
  end

  def self.silent?(message)
    message.split.join('').empty?
  end

  def self.respond(response)
    RESPONSES[response]
  end

  private_class_method :respond
end
